gen3qa.braincommons.org/dashboard/Secure/reports/dataHelper.js

```diff
@@ -0,0 +1,355 @@
+        )
+    };
+  }
+}
+
+
+/**
+ * Handler for result codes report
+ */
+export class RCodesHandler {
+  constructor() {
+  }
+
+  buildPathDateList() {
+    return basicBuildPathList("codes");
+  }
+
+  fetchData(pathDateList) {
+    return fetchBucketsSummary(pathDateList, "codes"
+      ).then(
+        (data) => {
+          return {
+            reportType: "rcodes",
+            data
+          };
+        }
+      );
+  }
+
+  massageData(fetchedData) {
+    return {
+      reportType: "rcodes",
+      massage: Object.entries(fetchedData.data).sort((a,b) => numCompare(a[0],b[0]))
+    };
+  }
+}
+
+/**
+ * Handler for result times report
+ */
+export class RTimesHandler {
+  constructor() {
+  }
+
+  buildPathDateList() {
+    return basicBuildPathList("rtimes");
+  }
+ 
+  fetchData(pathDateList) {
+    return fetchBucketsSummary(pathDateList, "rtimes"
+      ).then(
+        (data) => {
+          return {
+            reportType: "rtimes",
+            data
+          };
+        }
+      );
+  }
+
+  massageData(fetchedData) {
+    return {
+      reportType: "rtimes",
+      massage: Object.entries(fetchedData.data).sort((a,b) => numCompare(a[0],b[0]))
+    };
+  }
+}
+
+//----------------------------
+
+/**
+ * Generate mock data for testing UX without backend
+ */
+function fetchRecentDataMock() {
+  const result = {
+    users: {
+      reportType: "users",
+      data: [
+
+      ]
+    },
+    rcodes: {
+      reportType: "rcodes",
+      data: [
+
+      ]
+    },
+    rtimes: {
+      reportType: "rtimes",
+      data: [
+
+      ]
+    }
+  };
+
+  Object.keys(result).map(
+    (k) => {
+      let bar='*';
+      for(let i=0; i < 5; ++i) {
+        result[k].data.push(
+          [ k, "" + i, bar ] 
+        );
+        bar += '*';
+      }
+    }
+  );
+  return Promise.resolve(result);
+}
+
+
+const reportHandlers = {
+  rtimes: new RTimesHandler(),
+  rcodes: new RCodesHandler(),
+  users: new UniqueUsersHandler()
+};
+
+
+
+export function fetchRecentData(reportType) {
+  const handler = reportHandlers[reportType];
+  return handler.fetchData(handler.buildPathDateList()
+    ).then(
+      (data) => {
+        const result = handler.massageData(data);
+        result.data = data.data;
+        addStarColumn(result.massage);
+        return result;
+      }
+    );
+}
+
+export function amap(list, lambda, result=[]) {
+  if (list.length > 0) {
+    return lambda(list[0]).then(
+      (it) => {
+        result.push(it);
+        return amap(list.slice(1), lambda, result);
+      }
+    )
+  } else {
+    return Promise.resolve(result);
+  }
+}
+
+
+/**
+ * Try to retrieve data for the last 5 days
+ */
+function fetchRecentDataFromNetwork() {
+  const pathDateData = buildPathDateList();
+  const rawData = { users: [], rtimes: [], rcodes: [] };
+  fetchUserData(pathDateData.users
+  ).then(
+    function(userInfoList) {
+      rawData.users = userInfoList;
+      return fetchBucketsSummary(pathDateData.rcodes, );
```

Comment:
Fwiw fetchBucketsSummary seems to be missing a parameter here

Reply:
ugh!  good catch.  that's actually dead code - I removed it.  how about now?